src/App.tsx

```diff
@@ -1,26 +1,146 @@
+
+    todoService
+      .getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+        setTodosQuantity(
+          todosFromServer.filter(todo => !todo.completed).length,
+        );
+      })
+      .catch(err => {
+        setErrorMessage('Unable to load todos');
+        throw err;
+      });
+  }, []);
+
+  function addTodo(newTodo: Todo) {
+    setErrorMessage('');
+    setProcessing(0);
+    setTodos(currentTodos => [...currentTodos, newTodo]);
+
+    return todoService
+      .createTodo(newTodo)
+      .then(todo => setTodos(currentTodos => [...currentTodos, todo]))
+      .then(() =>
+        setTodos(currentTodos => {
+          currentTodos.splice(currentTodos.length - 2, 1);
+
+          if (!newTodo.completed) {
+            setTodosQuantity(todosQuantity + 1);
+          }
+
+          return currentTodos;
+        }),
+      )
+      .catch(err => {
+        setErrorMessage('Unable to add a todo');
+        setTodos(currentTodos => {
+          currentTodos.splice(currentTodos.length - 1, 1);
+
+          return currentTodos;
+        });
+        throw err;
+      })
+      .finally(() => setProcessing(undefined));
+  }
+
+  function deleteTodo(todoId: number) {
+    setErrorMessage('');
+    setProcessing(todoId);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then(() =>
+        setTodos(currentTodos => {
+          const foundedTodo = currentTodos.find(todo => todoId === todo.id);
+
+          if (!foundedTodo?.completed) {
+            setTodosQuantity(todosQuantity - 1);
+          }
+
+          return currentTodos.filter(todo => todo.id !== todoId);
+        }),
+      )
+      .catch(err => {
+        setErrorMessage('Unable to delete a todo');
+        throw err;
+      })
+      .finally(() => setProcessing(undefined));
+  }
+
+  const viewedTodos = filterTodos(todos, filter);
+
+  if (!todoService.USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">
-          React Todo App - Load Todos
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <TodoHeader
+          todos={todos}
+          onAdd={addTodo}
+          setError={setErrorMessage}
+          processing={processing}
+        />
+
+        <section className="todoapp__main" data-cy="TodoList">
+          {viewedTodos.map(todo => (
+            <TodoMain
+              todo={todo}
+              onDelete={deleteTodo}
+              processing={processing}
+              key={todo.id}
+            />
+          ))}
+        </section>
+
+        {todos.length !== 0 && (
```

Comment:
the same, but more readable
```suggestion
        {!!todos.length && (
```